crates/solidity/testing/sourcify/src/main.rs

```diff
@@ -0,0 +1,118 @@
+mod chains;
+mod command;
+mod compilation_builder;
+mod events;
+mod import_resolver;
+mod reporting;
+mod results;
+mod sourcify;
+mod tests;
+
+use std::path::PathBuf;
+
+use anyhow::Result;
+use clap::Parser;
+use command::{Commands, ShowCombinedResultsCommand};
+use events::Events;
+use infra_utils::github::GitHub;
+use infra_utils::paths::PathExtensions;
+use infra_utils::terminal::Terminal;
+use results::{display_all_results, AllResults};
+use sourcify::{ContractArchive, Manifest};
+use tests::{run_in_parallel, run_with_trace, test_single_contract};
+
+fn main() -> Result<()> {
+    let command::Cli { command } = command::Cli::parse();
+    match command {
+        Commands::Test(test_command) => run_test_command(test_command),
+        Commands::ShowCombinedResults(results_command) => {
+            run_show_combined_results_command(results_command)
+        }
+    }
+}
+
+// `cmd` is passed by value because it's consumed in the spawned thread,
+// but for whatever reason clippy can't figure that out
+#[allow(clippy::needless_pass_by_value)]
+fn run_test_command(cmd: command::TestCommand) -> Result<()> {
+    Terminal::step(format!(
+        "Initialize {chain}/{network}",
+        chain = cmd.chain.name(),
+        network = cmd.chain.network_name()
+    ));
+
+    let manifest = Manifest::new(cmd.chain, &cmd.sharding_options)?;
+
+    if let Some(contract) = &cmd.contract {
+        return test_single_contract(&manifest, contract, &cmd.test_options);
+    }
+
+    let archive_count = manifest.archive_count();
+
+    let (tx, rx) = std::sync::mpsc::channel::<ContractArchive>();
+
+    // Test archives which have been fetched and unpacked
+    let testing_thread = std::thread::spawn(move || -> Events {
+        let mut events = Events::new(archive_count, 0);
+        for archive in rx {
+            Terminal::step(archive.display_path());
+
+            events.start_archive(archive.contract_count());
+            if cmd.trace {
+                run_with_trace(&archive, &events, &cmd.test_options);
+            } else {
+                run_in_parallel(&archive, &events, &cmd.test_options);
+            }
+            events.finish_archive();
+
+            if !cmd.save {
+                archive.clean();
+            }
+        }
+
+        events
+    });
+
+    // Fetching the shards in this closure so that it takes ownership of the sender
+    // The sender needs to be dropped so that process_thread can finish
+    let fetcher = |t: std::sync::mpsc::Sender<ContractArchive>| {
+        for archive_desc in manifest.archives() {
+            let Ok(archive) = ContractArchive::fetch(archive_desc) else {
+                continue;
```

Comment:
Should we somehow indicate the I/O error? Otherwise it's possible the run succeeds, but in reality it didn't run any tests.